### PR TITLE
Add reference documentation for resolver settings

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

First part of https://github.com/astral-sh/uv/issues/5093.

Remaining:

- Global settings
- `pip`-specific settings (some will be copied-over from here)
- Auto-generating the "Possible values" for enums

### CODE REVIEW

File: crates/uv-settings/src/settings.rs [hunk 1]
```diff
@@ -136,24 +136,237 @@ pub struct ResolverInstallerOptions {
...
+    /// By default, `uv` will stop at the first index on which a given package is available, and
+    /// limit resolutions to those present on that first index (`first-match`). This prevents
+    /// "dependency confusion" attacks, whereby an attack can upload a malicious package under the
+    /// same name to a secondary.
+    ///
+    /// Possible values:
+    ///
+    /// - `"first-index"`:        Only use results from the first index that returns a match for a given package name.
+    /// - `"unsafe-first-match"`: Search for every package name across all indexes, exhausting the versions from the first index before moving on to the next.
+    /// - `"unsafe-best-match"`:  Search for every package name across all indexes, preferring the "best" version found. If a package version is in multiple indexes, only look at the entry for the first index.
```
Comment (charliermarsh) [inline crates/uv-settings/src/settings.rs:191]: Stopped adding these because I'm going to explore auto-generating them in the macro (separately).

File: crates/uv-cli/src/lib.rs [hunk 2]
```diff
@@ -2270,23 +2270,25 @@ pub struct IndexArgs {
     /// Accepts either a repository compliant with PEP 503 (the simple repository API), or a local
     /// directory laid out in the same format.
     ///
-    /// All indexes given via this flag take priority over the index
-    /// in `--index-url` (which defaults to PyPI). And when multiple
-    /// `--extra-index-url` flags are given, earlier values take priority.
+    /// All indexes provided via this flag take priority over the index specified by
+    /// `--index-url` (which defaults to PyPI). When multiple `--extra-index-url` flags are
+    /// provided, earlier values take priority.
```
Comment (zanieb) [inline crates/uv-cli/src/lib.rs:2275]: Should we reference the index strategy here?

File: crates/uv-settings/src/settings.rs [hunk 3]
```diff
@@ -44,7 +44,7 @@ pub struct Options {
         feature = "schemars",
         schemars(
             with = "Option<Vec<String>>",
-            description = "PEP 508 style requirements, e.g. `flask==3.0.0`, or `black @ https://...`."
+            description = "PEP 508 style requirements, e.g. `fastapi==3.0.0`, or `black @ https://...`."
```
Comment (zanieb) [inline crates/uv-settings/src/settings.rs:47]: Just wondering, why this change? Should we reference `ruff` instead?

Comment (charliermarsh) [reply crates/uv-settings/src/settings.rs:47]: I dunno, just thought I would use something a little more modern, but we can do whatever. Ruff seems good yeah.

File: crates/uv-settings/src/settings.rs [hunk 4]
```diff
@@ -136,24 +136,237 @@ pub struct ResolverInstallerOptions {
...
+            index-strategy = "unsafe-best-match"
+        "#
+    )]
     pub index_strategy: Option<IndexStrategy>,
+    /// Attempt to use `keyring` for authentication for index URLs.
+    ///
+    /// At present, only `--keyring-provider subprocess` is supported, which configures `uv` to
+    /// use the `keyring` CLI to handle authentication.
+    ///
+    /// Defaults to `disabled`.
```
Comment (zanieb) [inline crates/uv-settings/src/settings.rs:205]: Do we need this if covered in the `option` macro?

Comment (zanieb) [reply crates/uv-settings/src/settings.rs:205]: per https://github.com/astral-sh/uv/pull/5122/files#diff-cd5ab0390b5750a658337173d28e15c9631a156470fb8036633f3fefc4bf96a4R225-R227 this looks redundant

**Review by zanieb — APPROVED** [review]

File: crates/uv-settings/src/settings.rs (hunk 3 above)
Comment (zanieb) [reply crates/uv-settings/src/settings.rs:47]: I like to use Ruff when it's okay to have something that's not tiny.
